[master]Add Ballerina Deployment Guide (#9660)
ballerina-platform/ballerina-dev-website: swan-lake/development-tutorials/deployment-guide/serverless-deployment.md

---

1. [AWS Lambda](#aws-lambda)
2. [Azure Functions](#azure-functions)

## AWS Lambda

In this guide, we'll walk through deploying a serverless application using AWS Lambda with Ballerina. You'll learn how to set up, deploy, and manage your application, simplifying your production workflow and eliminating the need to manage servers.
An AWS Lambda function can be triggered by various AWS services. 

This example demonstrates how to write a simple echo function in AWS Lambda.

1. Set up the prerequisites. 
   For the instructions, see [Set up the prerequisites](https://ballerina.io/learn/aws-lambda/#set-up-the-prerequisites)
2. Here is an example of a Ballerina code for the function.
```ballerina
import ballerina/io;
import ballerinax/aws.lambda;

// The `@lambda:Function` annotation marks a function to generate an AWS Lambda function.
@lambda:Function
public function echo(lambda:Context ctx, json input) returns json {
    io:println(input.toJsonString());
    return input;
}
```

> 3. Use `bal build` for generating the AWS Lambda artifacts.

Reviewer: ```suggestion
3. Use `bal build` to generate the AWS Lambda artifacts.
```

The AWS Lambda functionality is implemented as a compiler extension. Therefore, artifact generation happens automatically when you build a Ballerina module by executing the command below.
```
$ bal build --cloud="aws_lambda"
```

> 4. Execute the AWS CLI command to deploy the function, here it creates and publishes the functions by replacing the respective AWS, $LAMBDA_ROLE_ARN, $REGION_ID, and $FUNCTION_NAME values given in the command with your values.

Reviewer: consider rewriting the sentence to avoid `, here`

- **Step 02:** This command tells Ballerina to compile the project specifically for AWS Lambda. It optimizes the output for Lambda’s serverless runtime requirements.
- **Step 03:** Package the compiled Ballerina application into a ZIP file suitable for AWS Lambda deployment.
- **Step 04:** Ensures that the workflow has the necessary credentials to interact with AWS services.
- **Step 05:** Deploy the packaged Ballerina function to AWS Lambda by uploading the ZIP file.
- **Step 06:** Test the deployed Lambda function by invoking it and outputting the response to confirm that the function works as expected.

## Azure Functions

In this guide, we'll walk through deploying a serverless application using Azure Functions with Ballerina. 
You'll learn how to set up, deploy, and manage your application, leveraging the power of Azure's serverless platform 
to simplify your workflows and eliminate the overhead of server management.

This example demonstrates how to write a simple echo function in Azure Functions using Ballerina, where triggers 
are represented by listeners attaching an `af:HttpListener` to a service that implies an HTTP trigger. 

> The resource method behaves like a `ballerina/http` service, In the provided code sample, there's an empty service path and a resource path

Reviewer: Are we starting a new sentence from `In the`?

named `hello` with a `get` accessor expecting a `name` query parameter, and the `ballerinax/azure_functions` package automatically handles the required artifact generation and data binding.

1. Set up the prerequisites by following [these instructions.](https://ballerina.io/learn/azure-functions/#set-up-the-prerequisites)

> 2. Here is an example Ballerina code.

Reviewer: ```suggestion
2. Here is an sample Ballerina code.
```

      # Step 4: Deploy to Azure Functions
      - name: Azure Functions Action
        uses: Azure/functions-action@v1
        with:
          app-name: 'your-function-app-name'
          package: './'
        env:
          AZURE_FUNCTIONAPP_PUBLISH_PROFILE: ${{ secrets.AZURE_FUNCTIONAPP_PUBLISH_PROFILE }}

      # Optional: Test the Azure Function
      - name: Invoke Azure Function
        run: |
          curl https://your-function-app-name.azurewebsites.net/api/your-function-name
```
- **Step 01:** The push event can initiate a build and deployment process
- **Step 02:** This step compiles the Ballerina application to generate an executable JAR file. The `bal build` command compiles the code and packages it, making it ready for deployment.
- **Step 03:** Install the Azure Functions Core Tools, which are necessary for deploying and testing Azure Functions in this GitHub Actions environment.
- **Step 04:** Deploy the built Ballerina application to an Azure Function App. This step uploads the packaged function to Azure and configures it for serverless execution.

For more reference,
- See [AWS Lambda](https://ballerina.io/learn/aws-lambda/) for more information.

> - See [Azure Functions](https://ballerina.io/learn/azure-functions/) for information about deployment options that can be used in Azure Functions.

Reviewer: Are we going to have above two pages in the learn section after these changes? If so what is the landing page location? And how do we handle the information loss (i.e. I see more information on those pages like supported triggers etc)

#### Build locally

The example command below will build the package and run it locally.

```
$ bal build --cloud="azure_functions_local" --graalvm
```

>**Info:** This will use the GraalVM you installed on your machine to build the native image and make the generated executable compatible with your machine.

#### Build for the cloud

The example command below will build the package for the Azure Functions cloud.

```
$ bal build --cloud="azure_functions" --graalvm
```

>**Info:** This will perform the compilation inside the Docker image to make it compatible with the Azure Functions cloud environment.

### Deploy the function

You can either deploy the Azure Functions package locally or on the Azure Functions cloud.

#### Deploy locally

The example command below will deploy the package locally.

```
$ func start --script-root target/azure_functions
```

#### Deploy on the cloud

The example command below will deploy the package on the Azure Functions cloud.

```
$ func azure functionapp publish <function_app_name> --script-root target/azure_functions
```

> ### CI/CD with Azure Function

Reviewer: ```suggestion
### CI/CD with Azure Functions
```

> ---

Reviewer: Check all the subtitles, seems the title levels are not correct.

<img width="263" alt="image" src="https://github.com/user-attachments/assets/a5e27a07-60be-40c1-9e1d-358345259fa4" />

To deploy the function, execute the command, which you get in the CLI output logs after you [build the function](#build-the-function). For examples, see [Examples](#examples).

>**Note:** When you are deploying, make sure to replace the `$FUNCTION_NAME`, `$LAMBDA_ROLE_ARN`, and `$REGION_ID` placeholders with the corresponding values you obtained when [setting up the prerequisites](#set-up-the-prerequisites).

>**Info:** For the supported parameters, go to the <a href="https://docs.aws.amazon.com/cli/latest/reference/lambda/create-function.html" target="_blank">`create-function` documentation</a>. You might need to change parameters such as the `MemorySize` and `Timeout` depending on your application and connection speed.

> ### Examples

Reviewer: Add the examples section as the last section under AWS Lambda, same applies to Azure functions section as well.
